File: app/workers/inference_worker.py
import numpy as np
from PySide6.QtCore import QThread, Signal
from app.ai.onnx_engine import ONNXModelEngine
from app.ai.color_key_engine import ColorKeyEngine
from app.utils.logger import logger

class BackgroundRemovalWorker(QThread):
    """
    Asynchronous worker thread to execute AI background removal
    without freezing the PySide6 main GUI thread.
    """
    progress = Signal(int)           # 0 to 100
    finished = Signal(np.ndarray)    # Emits generated alpha mask
    error = Signal(str)              # Emits error message string

    def __init__(self, image: np.ndarray, engine_type: str = "AI", model_name: str = "RMBG-1.4", device: str = "Auto", key_color=None, tolerance=40):
        super().__init__()
        self.image = image
        self.engine_type = engine_type  # "AI" or "ColorKey"
        self.model_name = model_name
        self.device = device
        self.key_color = key_color
        self.tolerance = tolerance

    def run(self):
        try:
            self.progress.emit(10)

            if self.engine_type == "ColorKey":
                engine = ColorKeyEngine()
                self.progress.emit(50)
                mask = engine.process(self.image, key_color=self.key_color, tolerance=self.tolerance)
                self.progress.emit(100)
                self.finished.emit(mask)
                return

            # AI Engine
            engine = ONNXModelEngine()
            self.progress.emit(25)

            if not engine.is_loaded() or engine._current_model_name != self.model_name:
                loaded = engine.load(model_name=self.model_name, device=self.device)
                if not loaded:
                    logger.warning("Could not load AI ONNX model. Falling back to classical segmentation.")

            self.progress.emit(60)
            mask = engine.process(self.image)
            self.progress.emit(100)

            self.finished.emit(mask)
        except Exception as e:
            logger.error(f"Error in background removal worker: {e}", exc_info=True)
            self.error.emit(str(e))

File: Cpp/src/workers/InferenceWorker.cpp
#include "workers/InferenceWorker.h"
#include "ai/ONNXEngine.h"
#include "ai/ColorKeyEngine.h"
#include "utils/Logger.h"

namespace ImageCut {
namespace Workers {

BackgroundRemovalWorker::BackgroundRemovalWorker(
    const cv::Mat& image,
    const QString& engineType,
    const QString& modelName,
    const QString& device,
    const QColor& keyColor,
    int tolerance
) : m_image(image.clone()),
    m_engineType(engineType),
    m_modelName(modelName),
    m_device(device),
    m_keyColor(keyColor),
    m_tolerance(tolerance)
{}

void BackgroundRemovalWorker::run() {
    LOG_INFO(QString("[WorkerThread] BackgroundRemovalWorker started. Engine: %1, Model: %2, Device: %3, Image size: %4x%5")
        .arg(m_engineType).arg(m_modelName).arg(m_device).arg(m_image.cols).arg(m_image.rows).toStdString());
    try {
        emit progress(10);

        if (m_engineType == "ColorKey") {
            LOG_INFO("[WorkerThread] Running ColorKeyEngine...");
            AI::ColorKeyEngine engine;
            emit progress(50);
            cv::Mat mask = engine.processColorKey(m_image, m_keyColor, m_tolerance);
            emit progress(100);
            LOG_INFO("[WorkerThread] ColorKeyEngine complete.");
            emit resultReady(mask);
            return;
        }

        LOG_INFO("[WorkerThread] Initializing ONNXModelEngine...");
        AI::ONNXModelEngine engine;
        emit progress(25);

        if (!engine.isLoaded()) {
            LOG_INFO("[WorkerThread] Loading ONNX model: " + m_modelName.toStdString());
            engine.load(m_modelName, m_device);
        }

        emit progress(60);
        LOG_INFO("[WorkerThread] Processing image through ONNXEngine...");
        cv::Mat mask = engine.process(m_image);
        emit progress(100);

        LOG_INFO(QString("[WorkerThread] Background removal finished successfully. Mask size: %1x%2")
            .arg(mask.cols).arg(mask.rows).toStdString());
        emit resultReady(mask);
    } catch (const std::exception& e) {
        LOG_ERROR(std::string("[WorkerThread] Exception caught: ") + e.what());
        emit error(QString::fromStdString(e.what()));
    } catch (...) {
        LOG_ERROR("[WorkerThread] Unknown exception caught during execution.");
        emit error("Unknown error occurred during background removal.");
    }
}

} // namespace Workers
} // namespace ImageCut
